File: java/sqlbuilder/Join.java
package com.clickability.cms.dataaccess.sqlbuilder;

import java.util.ArrayList;
import java.util.List;

public class Join implements Joinable {
	
	public enum Type {
		INNER("inner join"), OUTER("left outer join"),
		RIGHT_INNER("right inner join"), RIGHT_OUTER("right outer join");
		public final String name;
		private Type(String name) {
			this.name = name;
		}
	}

	private List<JoinExpression> joins;
	private Joinable base;

	public Join(Joinable base) {
		this.base = base;
		this.joins = new ArrayList<JoinExpression>();
	}
	
	public Join join(JoinExpression...expressions) {
		for (JoinExpression expr : expressions) {
			this.joins.add(expr);
		}
		return this;
	}
	
	@Override
	public String get() {
		String join = String.format("(%s)", this.base.get());
		for (JoinExpression expr: this.joins) {
			join += " " + expr.get();
		}
		return join;
	}
	
}

File: java/queryBuilder/Join.kt
package com.clickability.cms.dataaccess.sqlbuilder

import java.util.ArrayList

class Join(private val base: Joinable) : Joinable {

    private val joins: List<JoinExpression>

    enum class Type private constructor(val name: String) {
        INNER("inner join"), OUTER("left outer join"),
        RIGHT_INNER("right inner join"), RIGHT_OUTER("right outer join")
    }

    init {
        this.joins = ArrayList<JoinExpression>()
    }

    fun join(vararg expressions: JoinExpression): Join {
        for (expr in expressions) {
            this.joins.add(expr)
        }
        return this
    }

    @Override
    fun get(): String {
        var join = String.format("(%s)", this.base.get())
        for (expr in this.joins) {
            join += " " + expr.get()
        }
        return join
    }

}
